src/Test/DepTyCheck/Gen.idr

```diff
@@ -111,74 +150,149 @@ empty = Empty
+mapTaggedLazyElem xs f = mapElem xs $ \x, e => (fst x, delay $ f (snd x) $ elemMap _ e)
+
+mapOneOfElem : (gs : GenAlternatives ne iem a) ->
+               ((g : Gen iem a) -> (0 _ : Elem g gs) -> Gen em b) ->
+               GenAlternatives ne em b
+mapOneOfElem $ MkGenAlts gs = MkGenAlts . mapTaggedLazyElem gs
+
+0 allMapElem : {xs : LazyLst ne a} ->
+               {0 f : (x : a) -> (0 _ : Elem x xs) -> b} ->
+               ({x : a} -> (0 e : Elem x xs) -> p $ f x e) ->
+               All p $ mapElem xs f
+allMapElem {xs=[]}      _ = []
+allMapElem {xs=x :: xs} h = h Here :: allMapElem (\e => h $ There e)
+
+0 allMapTaggedLazyElem : {xs : LazyLst ne (tag, Lazy a)} ->
+                         {0 f : (x : a) -> (0 _ : Elem x $ unpackTaggedLazy xs) -> b} ->
+                         ({x : a} -> (0 e : Elem x $ unpackTaggedLazy xs) -> p $ f x e) ->
+                         All p $ unpackTaggedLazy $ mapTaggedLazyElem xs f
+allMapTaggedLazyElem h = allMap $ allMapElem $ \e => h $ elemMap _ e
+
+0 allMapOneOfElem : {gs : GenAlternatives ne iem a} ->
+                    {0 f : (g : Gen iem a) -> (0 _ : Elem g gs) -> Gen em b} ->
+                    ({g : Gen iem a} -> (0 e : Elem g gs) -> p $ f g e) ->
+                    All p $ mapOneOfElem gs f
+allMapOneOfElem {gs=MkGenAlts _} = allMapTaggedLazyElem
 
 -----------------------------
 --- Emptiness tweakenings ---
 -----------------------------
 
 export
 relax : (0 _ : iem `NoWeaker` em) => Gen iem a -> Gen em a
-relax @{nw} Empty      = rewrite maybeEmptyIsMinimal nw in Empty
-relax $ Pure x         = Pure x
-relax $ Raw x          = Raw x
-relax $ OneOf @{wo} x  = OneOf @{transitive' wo %search} x
-relax $ Bind @{bo} x f = Bind @{bindToOuterRelax bo %search} x f
-relax $ Labelled l x   = label l $ relax x
+0 relaxNonEmpty : {g : Gen iem a} -> IsNonEmpty g => IsNonEmpty $ relax @{nw} g
 
-export
-strengthen : {em : _} -> Gen iem a -> Maybe $ Gen em a
+relax @{nw} Empty           = rewrite maybeEmptyIsMinimal nw in Empty
+relax $ Pure x              = Pure x
+relax $ Raw x               = Raw x
+relax $ OneOf @{ne} @{nw} x = OneOf @{ne} @{transitive' nw %search} x
+relax $ Bind @{nw} x f      = Bind @{transitive' nw %search} x f
+relax $ Labelled l x        = Labelled @{relaxNonEmpty} l $ relax x
 
-strengthen {em=MaybeEmpty} Empty = Just Empty
-strengthen {em=_}          Empty = Nothing
+relaxNonEmpty {g=Pure _}       = Refl
+relaxNonEmpty {g=Raw _}        = Refl
+relaxNonEmpty {g=OneOf _}      = Refl
+relaxNonEmpty {g=Bind _ _}     = Refl
+relaxNonEmpty {g=Labelled _ _} = Refl
 
-strengthen $ Pure x = Just $ Pure x
-strengthen $ Raw x  = Just $ Raw x
+--------------------
+--- More utility ---
+--------------------
 
-strengthen $ OneOf @{_} @{au} x with (decCanBeEmpty em)
-  _ | Yes _ = Just $ OneOf @{relaxAnyCanBeEmpty au} @{au} x
-  _ | No  _ = map OneOf $ trMOneOf x $ assert_total $ strengthen {em=NonEmpty}
+strengthen : Gen em a -> Maybe $ Gen em a
```

Comment:
This function seems to change semantics, thus, maybe, it should be renamed or something. Originally it was used to match original `em` with the required one, if possible. Not `em` is the same in the start and in the end